adam/learner.py

```diff
@@ -102,3 +106,77 @@ def describe(
             return immutabledict(((memorized_description, 1.0),))
         else:
             return immutabledict()
+
+
+@attrs
+class SubsetLanguageLearner(
+    Generic[PerceptionT, LinguisticDescriptionT],
+    LanguageLearner[PerceptionT, LinguisticDescriptionT],
+):
+    """
+    An implementation of `LanguageLearner` for subset learning based approach for single object detection.
+    """
+
+    _descriptions_to_properties: Dict[
+        LinguisticDescriptionT, Set[PropertyPerception]
+    ] = attrib(init=False, default=Factory(dict))
+
+    def observe(
+        self, learning_example: LearningExample[PerceptionT, LinguisticDescriptionT]
+    ) -> None:
+        if len(learning_example.perception.frames) != 1:
+            raise RuntimeError("Subset learner can only handle single frames for now")
+
+        is_new_description = True
+        observed_linguistic_description = learning_example.linguistic_description
+        if isinstance(
+            learning_example.perception.frames[0], DevelopmentalPrimitivePerceptionFrame
+        ):
+            observed_property_assertions = learning_example.perception.frames[
+                0
+            ].property_assertions
+        else:
+            raise RuntimeError("Cannot process perception type.")
+
+        for description in self._descriptions_to_properties.keys():
+            # If already observed, reduce the properties for that description to the common subset of
+            # the new observation and the previous observations
+            if (
+                observed_linguistic_description.as_token_sequence()
+                == description.as_token_sequence()
+            ):
+                self._descriptions_to_properties[
+                    description
+                ] = self._descriptions_to_properties[description].intersection(
```

Comment:
Does doing the simple intersection here work? I thought perceived objects were never equal to one another?

Reply:
With this implementation we ensure we get the object that has the matching description, and intersect the properties, which can be equal. Here is an example for the 
```
First set: {hasProperty(learner, is-learner), hasProperty(ball_0, inanimate[binary]), hasColor(ball_0, #6e5f13), hasProperty(learner, self-moving[binary]), hasColor(ball_0, #c40233), hasProperty(learner, animate[binary])}
Second set: {hasColor(ball_0, #f2003c), hasProperty(ball_0, inanimate[binary]), hasProperty(learner, is-learner), hasProperty(learner, animate[binary]), hasProperty(learner, self-moving[binary])}
```